From 7e1606a5e452fc974b9623ce5fb9616fb70161aa Mon Sep 17 00:00:00 2001
From: Pranoy Goru <gorumanip@ami.com>
Date: Mon, 11 Dec 2023 16:18:37 +0530
Subject: [PATCH 157/157] Added Redfish Support for BRCM PCIE Switch

Description:

Added Basic get Support for BRCM PCIE switch in redfish

Test Case:
1) Get : https://{{BMCIP}}/redfish/v1/Oem/PcieSwitch

Response
{
    "@odata.id": "/redfish/v1/Oem/PcieSwitch",
    "@odata.type": "#PcieSwitch.PcieSwitch",
    "Members": [
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/1"
        }
    ],
    "Members@odata.count": 2,
    "Name": "PCIE Switch Collections"
}

2) Get instance for PCIE Switch
https://{{BMCIP}}/redfish/v1/Oem/PcieSwitch/0

Response
{
    "@odata.id": "/redfish/v1/Oem/PcieSwitch/0",
    "@odata.type": "#PcieSwitch.PcieSwitch",
    "ComponentId": 0,
    "EnclosureWwidHigh": 0,
    "EnclosureWwidLow": 0,
    "FwVersion": [
        0,
        0,
        0,
        0
    ],
    "HandleType": 258,
    "LastOp": "",
    "LastOpStatus": "Success",
    "Manufacturer": "Broadcom",
    "Mode": "BASE MODE",
    "Model": "Atlas2",
    "Name": "PCIE Switch",
    "NumLanes": 144,
    "PciAddrBusNumber": 0,
    "PciAddrDeviceNumber": 0,
    "PciAddrFunctionNumber": 0,
    "PciDeviceId": 49200,
    "PciMptDeviceId": 0,
    "PciSSDeviceId": 0,
    "PciSSVendorId": 0,
    "PciVendorId": 0,
    "Port": {
        "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports"
    },
    "Ports": [
        0,
        16,
        20,
        24,
        26,
        28,
        32,
        48,
        64,
        80,
        96,
        112,
        128,
        132,
        136,
        140
    ],
    "ProductHandle": 1,
    "RevisionId": 176,
    "SASAddressHigh": 0,
    "SASAddressLow": 0,
    "SbrVersion": [],
    "SmApiVersion": 0,
    "SwitchMode": 1,
    "Temperature": "40 C"
}

3) Ports Count
Get https://{{BMCIP}}/redfish/v1/Oem/PcieSwitch/0/Ports
Response:
{
    "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports",
    "@odata.type": "#PcieSwitch.PcieSwitch",
    "Members": [
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/0"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/16"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/20"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/24"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/26"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/28"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/32"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/48"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/64"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/80"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/96"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/112"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/128"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/132"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/136"
        },
        {
            "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/140"
        }
    ],
    "Members@odata.count": 16,
    "Name": "PCIE Switch Ports"
}

4) Instance for Ports
Get https://{{BMCIP}}/redfish/v1/Oem/PcieSwitch/0/Ports/16
{
    "@odata.id": "/redfish/v1/Oem/PcieSwitch/0/Ports/16",
    "@odata.type": "#PcieSwitch.PcieSwitch",
    "ActiveLanes": 15,
    "BusPrimary": 0,
    "BusSecondary": 0,
    "BusSubordinate": 0,
    "GID": 16,
    "LinkModeType": 0,
    "MaxLinkSpeed": 5,
    "MaxLinkWidth": 4,
    "MaxPayloadSize": 0,
    "MaxPayloadSizeSupport": 4,
    "MaxReadRequestSize": 0,
    "Name": "PCIE Switch Ports instance",
    "NegotiatedLinkSpeed": 3,
    "NegotiatedLinkWidth": 4,
    "PortNumber": 16,
    "PortType": 0,
    "Reserved": 0,
    "Reserved1": 0,
    "isEnabled": 1
}

Signed-off-by: Pranoy Goru <gorumanip@ami.com>
---
 redfish-core/include/redfish.hpp              |   6 +
 redfish-core/lib/oem_pcie_switch.hpp          | 343 ++++++++++++++++++
 redfish-core/lib/service_root.hpp             |   6 +
 static/redfish/v1/$metadata/index.xml         |  14 +
 .../AMIPcieSwitch/AMIPcieSwitch.json          | 255 +++++++++++++
 .../AMIPcieSwitchCollection.json              | 100 +++++
 .../AMIPcieSwitchPort/AMIPcieSwitchPort.json  | 147 ++++++++
 .../AMIPcieSwitchPortCollection.json          |  99 +++++
 .../OemServiceRoot/OemServiceRoot.json        |  71 ++++
 .../v1/schema/AMIPcieSwitchCollection_v1.xml  |  72 ++++
 .../schema/AMIPcieSwitchPortCollection_v1.xml |  67 ++++
 .../v1/schema/AMIPcieSwitchPort_v1.xml        | 163 +++++++++
 static/redfish/v1/schema/AMIPcieSwitch_v1.xml | 222 ++++++++++++
 .../redfish/v1/schema/OemServiceRoot_v1.xml   |  21 +-
 14 files changed, 1585 insertions(+), 1 deletion(-)
 create mode 100644 redfish-core/lib/oem_pcie_switch.hpp
 create mode 100644 static/redfish/v1/JsonSchemas/AMIPcieSwitch/AMIPcieSwitch.json
 create mode 100644 static/redfish/v1/JsonSchemas/AMIPcieSwitchCollection/AMIPcieSwitchCollection.json
 create mode 100644 static/redfish/v1/JsonSchemas/AMIPcieSwitchPort/AMIPcieSwitchPort.json
 create mode 100644 static/redfish/v1/JsonSchemas/AMIPcieSwitchPortCollection/AMIPcieSwitchPortCollection.json
 create mode 100644 static/redfish/v1/JsonSchemas/OemServiceRoot/OemServiceRoot.json
 create mode 100644 static/redfish/v1/schema/AMIPcieSwitchCollection_v1.xml
 create mode 100644 static/redfish/v1/schema/AMIPcieSwitchPortCollection_v1.xml
 create mode 100644 static/redfish/v1/schema/AMIPcieSwitchPort_v1.xml
 create mode 100644 static/redfish/v1/schema/AMIPcieSwitch_v1.xml

diff --git a/redfish-core/include/redfish.hpp b/redfish-core/include/redfish.hpp
index 85af1bb3..4fc77a67 100644
--- a/redfish-core/include/redfish.hpp
+++ b/redfish-core/include/redfish.hpp
@@ -47,6 +47,7 @@
 #include "node-manager/power.hpp"
 #include "node-manager/throttling_status.hpp"
 #include "node-manager/triggers.hpp"
+#include "oem_pcie_switch.hpp"
 #include "pcie.hpp"
 #include "pcie_slots.hpp"
 #include "pef_service.hpp"
@@ -308,6 +309,11 @@ class RedfishService
         requestRoutesMeterStateData(app);
 
         requestRoutesPefService(app);
+
+        requestRoutesPcieSwitchCollection(app);
+        requestRoutesPcieSwitchInstanceCollection(app);
+        requestRoutesPcieSwitchPortsCollection(app);
+        requestRoutesPcieSwitchPortsInstanceCollection(app);
         // Note, this must be the last route registered
         requestRoutesRedfish(app);
     }
diff --git a/redfish-core/lib/oem_pcie_switch.hpp b/redfish-core/lib/oem_pcie_switch.hpp
new file mode 100644
index 00000000..e3ab0b4e
--- /dev/null
+++ b/redfish-core/lib/oem_pcie_switch.hpp
@@ -0,0 +1,343 @@
+/*
+// Copyright (c) 2023 AMI Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+#pragma once
+
+// #include "bmcweb_config.h"
+
+#include "bmcweb_config.h"
+
+#include "app.hpp"
+#include "dbus_utility.hpp"
+#include "generated/enums/drive.hpp"
+#include "generated/enums/protocol.hpp"
+#include "health.hpp"
+#include "http/logging.hpp"
+#include "human_sort.hpp"
+#include "multipart_parser.hpp"
+#include "openbmc_dbus_rest.hpp"
+#include "ossl_random.hpp"
+#include "query.hpp"
+#include "redfish_util.hpp"
+#include "registries/privilege_registry.hpp"
+#include "task.hpp"
+#include "utils/collection.hpp"
+#include "utils/dbus_utils.hpp"
+#include "utils/sw_utils.hpp"
+
+#include <boost/algorithm/string/case_conv.hpp>
+#include <boost/system/error_code.hpp>
+#include <boost/url/format.hpp>
+#include <sdbusplus/asio/property.hpp>
+#include <sdbusplus/bus/match.hpp>
+#include <sdbusplus/unpack_properties.hpp>
+#include <task.hpp>
+
+#include <array>
+#include <filesystem>
+#include <string>
+#include <string_view>
+
+namespace redfish
+{
+
+inline void handleOemPcieSwitchCollectionGet(
+    App& app, const crow::Request& req,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
+{
+    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
+    {
+        return;
+    }
+    asyncResp->res.jsonValue["@odata.type"] =
+        "#AMIPcieSwitchCollection.AMIPcieSwitchCollection";
+    asyncResp->res.jsonValue["@odata.id"] = "/redfish/v1/Oem/AMI/PcieSwitch";
+    asyncResp->res.jsonValue["Name"] = "PCIE Switch Collections";
+    nlohmann::json& switchArray = asyncResp->res.jsonValue["Members"];
+    switchArray = nlohmann::json::array();
+    auto& count = asyncResp->res.jsonValue["Members@odata.count"];
+    count = 0;
+
+    crow::connections::systemBus->async_method_call(
+        [asyncResp, &switchArray,
+         &count](const boost::system::error_code ec,
+                 const std::vector<std::string>& switchList) {
+        if (ec)
+        {
+            BMCWEB_LOG_ERROR("DBUS error: no matched iface \n");
+            return;
+        }
+        for (const std::string& objpath : switchList)
+        {
+            std::size_t lastPos = objpath.rfind('/');
+            if (lastPos == std::string::npos || (objpath.size() <= lastPos + 1))
+            {
+                BMCWEB_LOG_ERROR("Failed to find '/' in ", objpath);
+                continue;
+            }
+            switchArray.push_back(
+                {{"@odata.id", "/redfish/v1/Oem/AMI/PcieSwitch/" +
+                                   objpath.substr(lastPos + 1)}});
+        } // object path loop
+        count = switchArray.size();
+    },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetSubTreePaths",
+        "/com/ami/PcieSwMgmt", 0,
+        std::array<const char*, 1>{"com.ami.PcieSwMgmt.BrcmAtlas.Manager"});
+}
+
+inline void handleOemPcieSwitchInstanceGet(
+    App& app, const crow::Request& req,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& switchId)
+{
+    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
+    {
+        return;
+    }
+    asyncResp->res.jsonValue["@odata.type"] =
+        "#AMIPcieSwitch.v1_0_0.AMIPcieSwitch";
+    asyncResp->res.jsonValue["@odata.id"] = "/redfish/v1/Oem/AMI/PcieSwitch/" +
+                                            switchId;
+    asyncResp->res.jsonValue["Id"] = switchId;
+    asyncResp->res.jsonValue["Name"] = "PCIE Switch";
+    asyncResp->res.jsonValue["Port"]["@odata.id"] = boost::urls::format(
+        "/redfish/v1/Oem/AMI/PcieSwitch/{}/Ports", switchId);
+
+    const std::string& serviceName = "com.ami.PcieSwMgmt.BrcmAtlas" + switchId;
+    const std::string& objPath = "/com/ami/PcieSwMgmt/BrcmAtlas/" + switchId;
+    crow::connections::systemBus->async_method_call(
+        [asyncResp](
+            const boost::system::error_code ec2,
+            const std::vector<std::pair<
+                std::string, std::variant<std::string, std::vector<uint32_t>,
+                                          uint32_t, uint16_t>>>&
+                propertiesList) {
+        if (ec2)
+        {
+            return;
+        }
+        for (const std::pair<std::string,
+                             std::variant<std::string, std::vector<uint32_t>,
+                                          uint32_t, uint16_t>>& property :
+             propertiesList)
+        {
+            const std::string& propertyName = property.first;
+            if ((propertyName == "ProductHandle") ||
+                (propertyName == "HandleType") ||
+                (propertyName == "SASAddressLow") ||
+                (propertyName == "SASAddressHigh") ||
+                (propertyName == "EnclosureWwidLow") ||
+                (propertyName == "EnclosureWwidHigh") ||
+                (propertyName == "SmApiVersion") ||
+                (propertyName == "SwitchMode") || (propertyName == "NumLanes"))
+            {
+                const uint32_t* value = std::get_if<uint32_t>(&property.second);
+                if (value != nullptr)
+                {
+                    asyncResp->res.jsonValue[propertyName] = *value;
+                }
+            }
+            if ((propertyName == "ComponentId") ||
+                (propertyName == "RevisionId") ||
+                (propertyName == "PciDeviceId") ||
+                (propertyName == "PciVendorId") ||
+                (propertyName == "PciMptDeviceId") ||
+                (propertyName == "PciSSDeviceId") ||
+                (propertyName == "PciSSVendorId") ||
+                (propertyName == "PciAddrBusNumber") ||
+                (propertyName == "PciAddrDeviceNumber") ||
+                (propertyName == "PciAddrFunctionNumber"))
+            {
+                const uint16_t* value = std::get_if<uint16_t>(&property.second);
+                if (value != nullptr)
+                {
+                    asyncResp->res.jsonValue[propertyName] = *value;
+                }
+            }
+            if ((propertyName == "Manufacturer") || (propertyName == "Model") ||
+                (propertyName == "Mode") || (propertyName == "Temperature") ||
+                (propertyName == "LastOp") || (propertyName == "LastOpStatus"))
+            {
+                const std::string* value =
+                    std::get_if<std::string>(&property.second);
+                if (value != nullptr)
+                {
+                    asyncResp->res.jsonValue[propertyName] = *value;
+                }
+            }
+            if ((propertyName == "Ports") || (propertyName == "SbrVersion") ||
+                (propertyName == "FwVersion"))
+            {
+                const std::vector<uint32_t>* value =
+                    std::get_if<std::vector<uint32_t>>(&property.second);
+                if (value != nullptr)
+                {
+                    asyncResp->res.jsonValue[propertyName] = *value;
+                }
+            }
+        }
+    },
+        serviceName, objPath, "org.freedesktop.DBus.Properties", "GetAll",
+        "com.ami.PcieSwMgmt.BrcmAtlas.Manager");
+}
+
+inline void handleOemPcieSwitchPortCollectionGet(
+    App& app, const crow::Request& req,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& switchId)
+{
+    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
+    {
+        return;
+    }
+
+    asyncResp->res.jsonValue["@odata.type"] =
+        "#AMIPcieSwitchPortCollection.AMIPcieSwitchPortCollection";
+    asyncResp->res.jsonValue["@odata.id"] = "/redfish/v1/Oem/AMI/PcieSwitch/" +
+                                            switchId + "/Ports";
+    asyncResp->res.jsonValue["Name"] = "PCIE Switch Ports";
+
+    const std::string& serviceName = "com.ami.PcieSwMgmt.BrcmAtlas" + switchId;
+    const std::string& objPath = "/com/ami/PcieSwMgmt/BrcmAtlas/" + switchId;
+
+    crow::connections::systemBus->async_method_call(
+        [asyncResp,
+         switchId](const boost::system::error_code ec,
+                   const std::variant<std::vector<uint32_t>>& portList) {
+        if (ec)
+        {
+            BMCWEB_LOG_DEBUG("failed to get property Value  ", ec);
+            return;
+        }
+        BMCWEB_LOG_DEBUG("## insideeeeBefore Dbus call");
+        nlohmann::json& switchArray = asyncResp->res.jsonValue["Members"];
+        switchArray = nlohmann::json::array();
+        auto& count = asyncResp->res.jsonValue["Members@odata.count"];
+        count = 0;
+
+        if (auto* vec = std::get_if<std::vector<uint32_t>>(&portList))
+        {
+            for (const auto& value : *vec)
+            {
+                switchArray.push_back(
+                    {{"@odata.id", "/redfish/v1/Oem/AMI/PcieSwitch/" +
+                                       switchId + "/Ports/" +
+                                       std::to_string(value)}});
+            }
+        }
+        count = switchArray.size();
+    },
+        serviceName, objPath, "org.freedesktop.DBus.Properties", "Get",
+        "com.ami.PcieSwMgmt.BrcmAtlas.Manager", "Ports");
+}
+
+inline void handleOemPcieSwitchPortInstanceGet(
+    App& app, const crow::Request& req,
+    const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+    const std::string& switchId, const std::string& portId)
+{
+    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
+    {
+        return;
+    }
+
+    asyncResp->res.jsonValue["@odata.type"] =
+        "#AMIPcieSwitchPort.v1_0_0.AMIPcieSwitchPort";
+    asyncResp->res.jsonValue["@odata.id"] = "/redfish/v1/Oem/AMI/PcieSwitch/" +
+                                            switchId + "/Ports/" + portId;
+    asyncResp->res.jsonValue["Name"] = "PCIE Switch Ports instance";
+    asyncResp->res.jsonValue["Id"] = portId;
+    const std::string& serviceName = "com.ami.PcieSwMgmt.BrcmAtlas" + switchId;
+    const std::string& objPath = "/com/ami/PcieSwMgmt/BrcmAtlas/" + switchId +
+                                 "/port/" + portId;
+    crow::connections::systemBus->async_method_call(
+        [asyncResp](
+            const boost::system::error_code ec2,
+            const std::vector<std::pair<std::string, std::variant<uint32_t>>>&
+                propertiesList) {
+        if (ec2)
+        {
+            return;
+        }
+        for (const std::pair<std::string, std::variant<uint32_t>>& property :
+             propertiesList)
+        {
+            const std::string& propertyName = property.first;
+            if ((propertyName == "PortType") ||
+                (propertyName == "PortNumber") ||
+                (propertyName == "isEnabled") || (propertyName == "GID") ||
+                (propertyName == "MaxReadRequestSize") ||
+                (propertyName == "MaxPayloadSize") ||
+                (propertyName == "MaxPayloadSizeSupport") ||
+                (propertyName == "NegotiatedLinkSpeed") ||
+                (propertyName == "MaxLinkSpeed") ||
+                (propertyName == "NegotiatedLinkWidth") ||
+                (propertyName == "MaxLinkWidth") ||
+                (propertyName == "LinkModeType") ||
+                (propertyName == "BusPrimary") ||
+                (propertyName == "BusSecondary") ||
+                (propertyName == "BusSubordinate") ||
+                (propertyName == "Reserved") ||
+                (propertyName == "ActiveLanes") ||
+                (propertyName == "Reserved1"))
+            {
+                const uint32_t* value = std::get_if<uint32_t>(&property.second);
+                if (value != nullptr)
+                {
+                    asyncResp->res.jsonValue[propertyName] = *value;
+                }
+            }
+        }
+    },
+        serviceName, objPath, "org.freedesktop.DBus.Properties", "GetAll",
+        "com.ami.PcieSwMgmt.BrcmAtlas.Ports.Manager");
+}
+
+inline void requestRoutesPcieSwitchCollection(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Oem/AMI/PcieSwitch")
+        .privileges({{"Login"}})
+        .methods(boost::beast::http::verb::get)(
+            std::bind_front(handleOemPcieSwitchCollectionGet, std::ref(app)));
+}
+
+inline void requestRoutesPcieSwitchInstanceCollection(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Oem/AMI/PcieSwitch/<str>")
+        .privileges({{"Login"}})
+        .methods(boost::beast::http::verb::get)(
+            std::bind_front(handleOemPcieSwitchInstanceGet, std::ref(app)));
+}
+
+inline void requestRoutesPcieSwitchPortsCollection(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Oem/AMI/PcieSwitch/<str>/Ports/")
+        .privileges({{"Login"}})
+        .methods(boost::beast::http::verb::get)(std::bind_front(
+            handleOemPcieSwitchPortCollectionGet, std::ref(app)));
+}
+
+inline void requestRoutesPcieSwitchPortsInstanceCollection(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Oem/AMI/PcieSwitch/<str>/Ports/<str>")
+        .privileges({{"Login"}})
+        .methods(boost::beast::http::verb::get)(
+            std::bind_front(handleOemPcieSwitchPortInstanceGet, std::ref(app)));
+}
+
+} // namespace redfish
diff --git a/redfish-core/lib/service_root.hpp b/redfish-core/lib/service_root.hpp
index 33339969..97403597 100644
--- a/redfish-core/lib/service_root.hpp
+++ b/redfish-core/lib/service_root.hpp
@@ -99,6 +99,12 @@ inline void handleServiceRootGetImpl(
         "/redfish/v1/#Oem/Intel";
     asyncResp->res.jsonValue["Oem"]["Intel"]["CupsService"] = {
         {"@odata.id", "/redfish/v1/Oem/Intel/CupsService"}};
+    asyncResp->res.jsonValue["Oem"]["AMI"]["@odata.type"] =
+        "#OemServiceRoot.v1_0_0.AMI";
+    asyncResp->res.jsonValue["Oem"]["AMI"]["@odata.id"] =
+        "/redfish/v1/#Oem/AMI";
+    asyncResp->res.jsonValue["Oem"]["AMI"]["PcieSwitch"] = {
+        {"@odata.id", "/redfish/v1/Oem/AMI/PcieSwitch"}};
 
     asyncResp->res.jsonValue["Links"]["ManagerProvidingService"]["@odata.id"] =
         "/redfish/v1/Managers/bmc";
diff --git a/static/redfish/v1/$metadata/index.xml b/static/redfish/v1/$metadata/index.xml
index 72397ad7..930f33f3 100644
--- a/static/redfish/v1/$metadata/index.xml
+++ b/static/redfish/v1/$metadata/index.xml
@@ -3468,4 +3468,18 @@
         <edmx:Include Namespace="OemDrive"/>
         <edmx:Include Namespace="OemDrive.v1_0_0"/>
     </edmx:Reference>
+    <edmx:Reference Uri="/redfish/v1/schema/AMIPcieSwitchCollection_v1.xml">
+        <edmx:Include Namespace="AMIPcieSwitchCollection"/>
+    </edmx:Reference>
+    <edmx:Reference Uri="/redfish/v1/schema/AMIPcieSwitch_v1.xml">
+        <edmx:Include Namespace="AMIPcieSwitch"/>
+	<edmx:Include Namespace="AMIPcieSwitch.v1_0_0"/>
+    </edmx:Reference>
+    <edmx:Reference Uri="/redfish/v1/schema/AMIPcieSwitchPortCollection_v1.xml">
+        <edmx:Include Namespace="AMIPcieSwitchPortCollection"/>
+   </edmx:Reference>
+   <edmx:Reference Uri="/redfish/v1/schema/AMIPcieSwitchPort_v1.xml">
+        <edmx:Include Namespace="AMIPcieSwitchPort"/>
+        <edmx:Include Namespace="AMIPcieSwitchPort.v1_0_0"/>
+    </edmx:Reference>
 </edmx:Edmx>
diff --git a/static/redfish/v1/JsonSchemas/AMIPcieSwitch/AMIPcieSwitch.json b/static/redfish/v1/JsonSchemas/AMIPcieSwitch/AMIPcieSwitch.json
new file mode 100644
index 00000000..f193a387
--- /dev/null
+++ b/static/redfish/v1/JsonSchemas/AMIPcieSwitch/AMIPcieSwitch.json
@@ -0,0 +1,255 @@
+{
+    "$id": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitch.v1_0_0.json",
+    "$schema": "http://redfish.dmtf.org/schemas/v1/redfish-schema-v1.json",
+    "copyright": "Copyright 2014-2023 DMTF. For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright",
+    "definitions": {
+        "Actions": {
+            "additionalProperties": false,
+            "description": "The available actions for this resource.",
+            "longDescription": "This type shall contain the available actions for this resource.",
+            "patternProperties": {
+                "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                    "description": "This property shall specify a valid odata or Redfish property.",
+                    "type": [
+                        "array",
+                        "boolean",
+                        "integer",
+                        "number",
+                        "null",
+                        "object",
+                        "string"
+                    ]
+                }
+            },
+            "properties": {
+                "Oem": {
+                    "$ref": "#/definitions/OemActions",
+                    "description": "The available OEM-specific actions for this resource.",
+                    "longDescription": "This property shall contain the available OEM-specific actions for this resource."
+                }
+            },
+            "type": "object"
+        },
+        "AMIPcieSwitch": {
+            "additionalProperties": false,
+            "description": "The AMIPcieSwitch schema contains properties that describe a device details.",
+            "longDescription": "This resource contains a simple PcieSwitch for a Redfish implementation.",
+            "patternProperties": {
+                "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                    "description": "This property shall specify a valid odata or Redfish property.",
+                    "type": [
+                        "array",
+                        "boolean",
+                        "integer",
+                        "number",
+                        "null",
+                        "object",
+                        "string"
+                    ]
+                }
+            },
+            "properties": {
+                "@odata.context": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/context"
+                },
+                "@odata.etag": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/etag"
+                },
+                "@odata.id": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/id"
+                },
+                "@odata.type": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/type"
+                },
+                "Actions": {
+                    "$ref": "#/definitions/Actions",
+                    "description": "The available actions for this resource.",
+                    "longDescription": "This property shall contain the available actions for this resource."
+                },
+		"ComponentId": {
+                    "description": "Represents ComponentId.",
+                    "longDescription": "Represents ComponentId of an intance",
+                    "type": "number"
+                },
+                "EnclosureWwidHigh": {
+                    "description": "It indicates EnclosureWwidHigh.",
+                    "longDescription": "It indicates EnclosureWwidHigh.",
+                    "type": "number"
+                },
+		"EnclosureWwidLow": {
+                    "description": "It indicates EnclosureWwidLow.",
+                    "longDescription": "It indicates EnclosureWwidLow.",
+                    "type": "number"
+                },
+		"FwVersion": {
+                    "description": "It indicates FwVersion.",
+                    "longDescription": "It indicates FwVersion.",
+                    {
+                        "type": [
+                            "number",
+                            "null"
+                        ]
+                    },
+                    "type": "array"
+                },
+		"HandleType": {
+                    "description": "It indicates HandleType.",
+                    "longDescription": "It indicates HandleType.",
+                    "type": "number"
+                },
+		"LastOp": {
+                    "description": "It indicates LastOp.",
+                    "longDescription": "It indicates LastOp.",
+                    "type": [
+                        "string",
+                        "null"
+                    ]
+                },
+		"LastOpStatus": {
+                    "description": "It indicates LastOpStatus.",
+                    "longDescription": "It indicates LastOpStatus.",
+	                "type": [
+        	           "string",
+                           "null"
+                     ]
+		},
+		"Manufacturer": {
+                    "description": "It indicates Manufacturer.",
+                    "longDescription": "It indicates Manufacturer.",
+                    "type": [
+                        "string",
+                        "null"
+		    ]
+                },
+		"Mode": {
+                    "description": "It indicates Mode.",
+                    "longDescription": "It indicates Mode.",
+                    "type": [
+                        "string",
+                        "null"
+                    ]
+                },
+		"Model": {
+                    "description": "It indicates Model.",
+                    "longDescription": "It indicates Model.",
+                    "type": [
+                        "string",
+                        "null"
+                    ]
+                },
+		"Name": {
+                    "description": "It indicates Name.",
+                    "longDescription": "It indicates Name.",
+                    "type": [
+                        "string",
+                        "null"
+                    ]
+                },
+		"NumLanes": {
+                    "description": "It indicates NumLanes.",
+                    "longDescription": "It indicates NumberLanes.",
+                    "type": "number"
+                },
+		"PciAddrBusNumber": {
+                    "description": "It indicates PciAddrBusNumber.",
+                    "longDescription": "It indicates PciAddrBusNumber.",
+                    "type": "number"
+                },
+		"PciAddrDeviceNumber": {
+                    "description": "It indicates PciAddrDeviceNumber.",
+                    "longDescription": "It indicates PciAddrDeviceNumber.",
+                    "type": "number"
+                },
+		"PciAddrFunctionNumber": {
+                    "description": "It indicates PciAddrFunctionNumber.",
+                    "longDescription": "It indicates PciAddrFunctionNumber.",
+                    "type": "number"
+                },
+		"PciDeviceId": {
+                    "description": "It indicates PciDeviceID.",
+                    "longDescription": "It indicates PciDeviceId.",
+                    "type": "number"
+                },
+		"PciMptDeviceId": {
+                    "description": "It indicates PciMptDeviceId.",
+                    "longDescription": "It indicates PciMptDeviceId.",
+                    "type": "number"
+                },
+		"PciSSDeviceId": {
+                    "description": "It indicates PciSSDeviceId.",
+                    "longDescription": "It indicates PciSSDeviceId.",
+                    "type": "number"
+                },
+		"PciSSVendorId": {
+                    "description": "It indicates PciSSVendorId.",
+                    "longDescription": "It indicates PciSSVendorId.",
+                    "type": "number"
+                },
+		"PciVendorId": {
+                    "description": "It indicates PciVendorId.",
+                    "longDescription": "It indicates PciVendorId.",
+                    "type": "number"
+                },
+		"Port": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchPortCollection.json#/definitions/AMIPcieSwitchPortCollection",
+                    "description": "The link to AMIPcieSwitchPortCollection",
+                    "longDescription": "This property shall contain a link to a Resource of type AMIPcieSwitchPort collection.",
+                    "readonly": true
+                },
+		"Ports": {
+                    "description": "It indicates Ports.",
+                    "longDescription": "It indicates Ports.",
+                    {
+                        "type": [
+                            "string",
+                            "null"
+                        ]
+                    },
+		    "type": "array"
+                },
+		"ProductHandle": {
+                    "description": "It indicates ProductHandle.",
+                    "longDescription": "It indicates ProductHandle.",
+                    "type": "number"
+                },
+		"RevisionId": {
+                    "description": "It indicates RevisionId.",
+                    "longDescription": "It indicates RevisionId.",
+                    "type": "number"
+                },
+		"SASAddressHigh": {
+                    "description": "It indicates SASAddressHigh.",
+                    "longDescription": "It indicates SASAddressHigh.",
+                    "type": "number"
+                },
+		"SASAddressLow": {
+                    "description": "It indicates SASAddressLow.",
+                    "longDescription": "It indicates SASAddressLow.",
+                    "type": "number"
+                },
+		"SbrVersion": {
+                    "description": "It indicates SbrVersion.",
+                    "longDescription": "It indicates SbrVersion.",
+                    "type": "number"
+                },
+		"SmApiVersion": {
+                    "description": "It indicates SmApiVersion.",
+                    "longDescription": "It indicates SmApiVersion.",
+                    "type": "number"
+                },
+		"SwitchMode": {
+                    "description": "It indicates SwitchMode.",
+                    "longDescription": "It indicates SwitchMode.",
+                    "type": "number"
+                },
+		"Temperature": {
+                    "description": "It indicates Temperature.",
+                    "longDescription": "It indicates Temperature.",
+                    "type": "number"
+                }
+	    },
+            "type": "object"
+        }
+    },
+    "title": "#AMIPcieSwitch"
+}
diff --git a/static/redfish/v1/JsonSchemas/AMIPcieSwitchCollection/AMIPcieSwitchCollection.json b/static/redfish/v1/JsonSchemas/AMIPcieSwitchCollection/AMIPcieSwitchCollection.json
new file mode 100644
index 00000000..af95c5f1
--- /dev/null
+++ b/static/redfish/v1/JsonSchemas/AMIPcieSwitchCollection/AMIPcieSwitchCollection.json
@@ -0,0 +1,100 @@
+{
+    "$id": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchCollection.json",
+    "$schema": "http://redfish.dmtf.org/schemas/v1/redfish-schema-v1.json",
+    "copyright": "Copyright 2014-2023 DMTF. For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright",
+    "definitions": {
+        "AMIPcieSwitchCollection": {
+            "anyOf": [
+                {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/idRef"
+                },
+                {
+                    "additionalProperties": false,
+                    "description": "The AMIPcieSwitchCollection schema describes a collection of PcieSwitch instances.",
+                    "longDescription": "This Resource shall represent a Resource Collection of PcieSwitch instances for a Redfish implementation.",
+                    "patternProperties": {
+                        "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                            "description": "This property shall specify a valid odata or Redfish property.",
+                            "type": [
+                                "array",
+                                "boolean",
+                                "integer",
+                                "number",
+                                "null",
+                                "object",
+                                "string"
+                            ]
+                        }
+                    },
+                    "properties": {
+                        "@odata.context": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/context"
+                        },
+                        "@odata.etag": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/etag"
+                        },
+                        "@odata.id": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/id"
+                        },
+                        "@odata.type": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/type"
+                        },
+                        "Description": {
+                            "anyOf": [
+                                {
+                                    "$ref": "http://redfish.dmtf.org/schemas/v1/Resource.json#/definitions/Description"
+                                },
+                                {
+                                    "type": "null"
+                                }
+                            ],
+                            "readonly": true
+                        },
+                        "Members": {
+                            "description": "The members of this collection.",
+                            "items": {
+                                "$ref": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitch.json#/definitions/AMIPcieSwitch"
+                            },
+                            "longDescription": "This property shall contain an array of links to the members of this collection.",
+                            "readonly": true,
+                            "type": "array"
+                        },
+                        "Members@odata.count": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/count"
+                        },
+                        "Members@odata.nextLink": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/nextLink"
+                        },
+                        "Name": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/Resource.json#/definitions/Name",
+                            "readonly": true
+                        },
+                        "Oem": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/Resource.json#/definitions/Oem",
+                            "description": "The OEM extension property.",
+                            "longDescription": "This property shall contain the OEM extensions.  All values for properties contained in this object shall conform to the Redfish Specification-described requirements."
+                        }
+                    },
+                    "required": [
+                        "Members",
+                        "Members@odata.count",
+                        "@odata.id",
+                        "@odata.type",
+                        "Name"
+                    ],
+                    "type": "object"
+                }
+            ],
+            "deletable": false,
+            "insertable": false,
+            "updatable": false,
+            "uris": [
+                "/redfish/v1/Oem/AMI/PcieSwitch"
+            ]
+        }
+    },
+    "owningEntity": "AMI",
+    "title": "#AMIPcieSwitchCollection.AMIPcieSwitchCollection"
+}
+~
+
diff --git a/static/redfish/v1/JsonSchemas/AMIPcieSwitchPort/AMIPcieSwitchPort.json b/static/redfish/v1/JsonSchemas/AMIPcieSwitchPort/AMIPcieSwitchPort.json
new file mode 100644
index 00000000..ae0a4cf1
--- /dev/null
+++ b/static/redfish/v1/JsonSchemas/AMIPcieSwitchPort/AMIPcieSwitchPort.json
@@ -0,0 +1,147 @@
+{
+    "$id": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchPort.v1_0_0.json",
+    "$schema": "http://redfish.dmtf.org/schemas/v1/redfish-schema-v1.json",
+    "copyright": "Copyright 2014-2023 DMTF. For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright",
+    "definitions": {
+        "AMIPcieSwitchPort": {
+            "additionalProperties": false,
+            "description": "The PcieSwitch shema contains PcieSwitch Port details.",
+            "longDescription": "The PcieSwitch shema contains PcieSwitch Port details.",
+            "patternProperties": {
+                "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                    "description": "This property shall specify a valid odata or Redfish property.",
+                    "type": [
+                        "array",
+                        "boolean",
+                        "integer",
+                        "number",
+                        "null",
+                        "object",
+                        "string"
+                    ]
+                }
+            },
+            "properties": {
+                "@odata.context": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/context"
+                },
+                "@odata.etag": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/etag"
+                },
+                "@odata.id": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/id"
+                },
+                "@odata.type": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/type"
+                },
+                "ActiveLanes": {
+                    "description": "Represents ActiveLanes.",
+                    "longDescription": "This property indicates ActiveLanes",
+                    "type": "number"
+                },
+                "BusPrimary": {
+                    "description": "This property indicates BusPrimary.",
+                    "longDescription": "This property indicates BusPrimary.",
+                    "type": "number"
+                },
+                "BusSecondary": {
+                    "description": "This property indicates BusSecondary",
+                    "longDescription": "This property indicates BusSecondary.",
+                    "type": "number"
+                },
+                "BusSubordinate": {
+                    "description": "Represents BusSubordinate.",
+                    "longDescription": "This property indicates BusSubordinate..",
+                    "type": "number"
+                },
+                "GID": {
+                    "description": "Represents GID.",
+                    "longDescription": "This property indicates GID.",
+                    "type": "number"
+                },
+		"Id": {
+                    "description": "This property indicates ID.",
+                    "longDescription": "This property indicates ID.",
+                    "type": "number"
+                },
+		"LinkModeType": {
+                    "description": "It indicates LinkModeType.",
+                    "longDescription": "It indicates LinkModeType.",
+                    "type": "number"
+                },
+		"MaxLinkSpeed": {
+                    "description": "It indicates MaxLinkSpeed.",
+                    "longDescription": "It indicates MaxLinkSpeed..",
+                    "type": "number"
+                },
+		"MaxLinkWidth": {
+                    "description": "It indicates MaxLinkWidth.",
+                    "longDescription": "It indicates MaxLinkWidth.",
+                    "type": "number"
+                },
+		"MaxPayloadSize": {
+                    "description": "It indicates MaxPayloadSize.",
+                    "longDescription": "It indicates MaxPayloadSize.",
+                    "type": "number"
+                },
+		"MaxPayloadSizeSupport": {
+                    "description": "It indicates MaxPayloadSizeSupport.",
+                    "longDescription": "It indicates MaxPayloadSizeSupport.",
+                    "type": "number"
+                },
+		"MaxReadRequestSize": {
+                    "description": "It indicates MaxReadRequestSize.",
+                    "longDescription": "It indicates MaxReadRequestSize.",
+                    "type": "number"
+                },
+		"NegotiatedLinkSpeed": {
+                    "description": "It indicates NegotiatedLinkSpeed.",
+                    "longDescription": "It indicates NegotiatedLinkSpeed.",
+                    "type": "number"
+                },
+		"NegotiatedLinkWidth": {
+                    "description": "It indicates NegotiatedLinkWidth.",
+                    "longDescription": "It indicates NegotiatedLinkWidth.",
+                    "type": "number"
+                },
+		"PortNumber": {
+                    "description": "It indicates PortNumber.",
+                    "longDescription": "It indicates PortNumber.",
+                    "type": "number"
+                },
+		"PortType": {
+                    "description": "It indicates PortType.",
+                    "longDescription": "It indicates PortType.",
+                    "type": "number"
+                },
+		"Reserved": {
+                    "description": "It indicates Reserved status.",
+                    "longDescription": "It indicates Reserved status.",
+                    "type": "number"
+                },
+		"Reserved1": {
+                    "description": "It indicates Reserved1 status.",
+                    "longDescription": "It indicates Reserved1 status.",
+                    "type": "number"
+                },
+		"isEnabled": {
+                    "description": "It indicates Port is Enabled or not.",
+                    "longDescription": "It indicates Port is Enabled or not.",
+                    "type": "number"
+                },
+                "Name": {
+                    "description": "It indicates Name of a Port.",
+                    "longDescription": "It indicates Name of a Port.",
+                    "type": [
+                        "string",
+                        "null"
+                    ]
+                }
+	    },
+            "type": "object"
+        }
+    },
+    "title": "#AMIPcieSwitchPort"
+}
+
+
diff --git a/static/redfish/v1/JsonSchemas/AMIPcieSwitchPortCollection/AMIPcieSwitchPortCollection.json b/static/redfish/v1/JsonSchemas/AMIPcieSwitchPortCollection/AMIPcieSwitchPortCollection.json
new file mode 100644
index 00000000..d8101293
--- /dev/null
+++ b/static/redfish/v1/JsonSchemas/AMIPcieSwitchPortCollection/AMIPcieSwitchPortCollection.json
@@ -0,0 +1,99 @@
+{
+    "$id": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchPortCollection.json",
+    "$schema": "http://redfish.dmtf.org/schemas/v1/redfish-schema-v1.json",
+    "copyright": "Copyright 2014-2023 DMTF. For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright",
+    "definitions": {
+        "AMIPcieSwitchPortCollection": {
+            "anyOf": [
+                {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/idRef"
+                },
+                {
+                    "additionalProperties": false,
+                    "description": "The AMIPcieSwitchPortCollection schema describes a collection of PcieSwitch Port instances.",
+                    "longDescription": "This Resource shall represent a Resource Collection of PcieSwitch Port instances for a Redfish implementation.",
+                    "patternProperties": {
+                        "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                            "description": "This property shall specify a valid odata or Redfish property.",
+                            "type": [
+                                "array",
+                                "boolean",
+                                "integer",
+                                "number",
+                                "null",
+                                "object",
+                                "string"
+                            ]
+                        }
+                    },
+                    "properties": {
+                        "@odata.context": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/context"
+                        },
+                        "@odata.etag": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/etag"
+                        },
+                        "@odata.id": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/id"
+                        },
+                        "@odata.type": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/type"
+                        },
+                        "Description": {
+                            "anyOf": [
+                                {
+                                    "$ref": "http://redfish.dmtf.org/schemas/v1/Resource.json#/definitions/Description"
+                                },
+                                {
+                                    "type": "null"
+                                }
+                            ],
+                            "readonly": true
+                        },
+                        "Members": {
+                            "description": "The members of this collection.",
+                            "items": {
+                                "$ref": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchPort.json#/definitions/AMIPcieSwitchPort"
+                            },
+                            "longDescription": "This property shall contain an array of links to the members of this collection.",
+                            "readonly": true,
+                            "type": "array"
+                        },
+                        "Members@odata.count": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/count"
+                        },
+                        "Members@odata.nextLink": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/odata-v4.json#/definitions/nextLink"
+                        },
+                        "Name": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/Resource.json#/definitions/Name",
+                            "readonly": true
+                        },
+                        "Oem": {
+                            "$ref": "http://redfish.dmtf.org/schemas/v1/Resource.json#/definitions/Oem",
+                            "description": "The OEM extension property.",
+                            "longDescription": "This property shall contain the OEM extensions.  All values for properties contained in this object shall conform to the Redfish Specification-described requirements."
+                        }
+                    },
+                    "required": [
+                        "Members",
+                        "Members@odata.count",
+                        "@odata.id",
+                        "@odata.type",
+                        "Name"
+                    ],
+                    "type": "object"
+                }
+            ],
+            "deletable": false,
+            "insertable": false,
+            "updatable": false,
+            "uris": [
+                "/redfish/v1/Oem/AMI/PcieSwitch/{}/Ports"
+            ]
+        }
+    },
+    "owningEntity": "AMI",
+    "title": "#AMIPcieSwitchPortCollection.AMIPcieSwitchPortCollection"
+}
+
diff --git a/static/redfish/v1/JsonSchemas/OemServiceRoot/OemServiceRoot.json b/static/redfish/v1/JsonSchemas/OemServiceRoot/OemServiceRoot.json
new file mode 100644
index 00000000..c42c80b9
--- /dev/null
+++ b/static/redfish/v1/JsonSchemas/OemServiceRoot/OemServiceRoot.json
@@ -0,0 +1,71 @@
+{
+    "$id": "http://redfish.dmtf.org/schemas/v1/OemServiceRoot.json",
+    "$schema": "http://redfish.dmtf.org/schemas/v1/redfish-schema-v1.json",
+    "copyright": "Copyright 2014-2019 DMTF. For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright",
+    "definitions": {
+        "AMI": {
+            "additionalProperties": false,
+            "description": "AMI Oem Extension for ServiceRoot",
+            "longDescription": "AMI Oem Extension for ServiceRoot",
+            "patternProperties": {
+                "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                    "description": "This property shall specify a valid odata or Redfish property.",
+                    "type": [
+                        "array",
+                        "boolean",
+                        "integer",
+                        "number",
+                        "null",
+                        "object",
+                        "string"
+                    ]
+                }
+            },
+            "properties": {
+		"@odata.context": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/odata.4.0.0.json#/definitions/context"
+                },
+                "PcieSwicth": {
+                    "$ref": "http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchCollection.json#/definitions/AMIPcieSwitchCollection",
+                    "description": "The link to AMIPcieSwitchCollection",
+                    "longDescription": "This property shall contain a link to a Resource of type AMIPcieSwitch collection.",
+                    "readonly": true
+                }
+            },
+            "type": "object"
+        },
+        "Oem": {
+            "additionalProperties": true,
+            "description": "OemUpdateService Oem properties.",
+            "patternProperties": {
+                "^([a-zA-Z_][a-zA-Z0-9_]*)?@(odata|Redfish|Message)\\.[a-zA-Z_][a-zA-Z0-9_]*$": {
+                    "description": "This property shall specify a valid odata or Redfish property.",
+                    "type": [
+                        "array",
+                        "boolean",
+                        "integer",
+                        "number",
+                        "null",
+                        "object",
+                        "string"
+                    ]
+                }
+            },
+            "properties": {
+                "AMI": {
+                    "anyOf": [
+                        {
+                            "$ref": "#/definitions/AMI"
+                        },
+                        {
+                            "type": "null"
+                        }
+                    ]
+                }
+            },
+            "type": "object"
+        }
+    },
+    "title": "#OemServiceRoot"
+}
+
diff --git a/static/redfish/v1/schema/AMIPcieSwitchCollection_v1.xml b/static/redfish/v1/schema/AMIPcieSwitchCollection_v1.xml
new file mode 100644
index 00000000..81599697
--- /dev/null
+++ b/static/redfish/v1/schema/AMIPcieSwitchCollection_v1.xml
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!---->
+<!--################################################################################       -->
+<!--# Redfish Schema:  AMIPcieSwitchCollection                                             -->
+<!--#                                                                                      -->
+<!--# For a detailed change log, see the README file contained in the DSP8010 bundle,      -->
+<!--# available at http://www.dmtf.org/standards/redfish                                   -->
+<!--# Copyright 2014-2020 DMTF.                                                            -->
+<!--# For the full DMTF copyright policy, see http://www.dmtf.org/about/policies/copyright -->
+<!--################################################################################       -->
+<!---->
+<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.0">
+
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Core.V1.xml">
+    <edmx:Include Namespace="Org.OData.Core.V1" Alias="OData"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Capabilities.V1.xml">
+    <edmx:Include Namespace="Org.OData.Capabilities.V1" Alias="Capabilities"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/Resource_v1.xml">
+    <edmx:Include Namespace="Resource.v1_0_0"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/RedfishExtensions_v1.xml">
+    <edmx:Include Namespace="RedfishExtensions.v1_0_0" Alias="Redfish"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/AMIPcieSwitch_v1.xml">
+    <edmx:Include Namespace="AMIPcieSwitch"/>
+  </edmx:Reference>
+
+  <edmx:DataServices>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="AMIPcieSwitchCollection">
+      <Annotation Term="Redfish.OwningEntity" String="AMI"/>
+
+      <EntityType Name="AMIPcieSwitchCollection" BaseType="Resource.v1_0_0.ResourceCollection">
+        <Annotation Term="OData.Description" String="The collection of AMIPcieSwitch Resource instances."/>
+        <Annotation Term="OData.LongDescription" String="This Resource shall represent a Resource Collection of AMIPcieSwitch instances for a Redfish implementation."/>
+        <Annotation Term="Capabilities.InsertRestrictions">
+          <Record>
+            <PropertyValue Property="Insertable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.UpdateRestrictions">
+          <Record>
+            <PropertyValue Property="Updatable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.DeleteRestrictions">
+          <Record>
+            <PropertyValue Property="Deletable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Redfish.Uris">
+          <Collection>
+		  <String>/redfish/v1/Oem/AMI/PcieSwitch</String>
+          </Collection>
+  </Annotation>
+
+        <NavigationProperty Name="Members" Type="Collection(AMIPcieSwitch.AMIPcieSwitch)">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/Read"/>
+          <Annotation Term="OData.Description" String="The members of this collection."/>
+          <Annotation Term="OData.LongDescription" String="This property shall contain an array of links to the members of this collection."/>
+          <Annotation Term="OData.AutoExpandReferences"/>
+          <Annotation Term="Redfish.Required"/>
+        </NavigationProperty>
+      </EntityType>
+
+    </Schema>
+
+  </edmx:DataServices>
+</edmx:Edmx>
+
diff --git a/static/redfish/v1/schema/AMIPcieSwitchPortCollection_v1.xml b/static/redfish/v1/schema/AMIPcieSwitchPortCollection_v1.xml
new file mode 100644
index 00000000..6fa8aae7
--- /dev/null
+++ b/static/redfish/v1/schema/AMIPcieSwitchPortCollection_v1.xml
@@ -0,0 +1,67 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!---->
+<!--################################################################################       -->
+<!--# Redfish Schema:  AMIPcieSwitchPortCollection                                         -->
+<!--#                                                                                      -->
+<!--################################################################################       -->
+<!---->
+<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.0">
+
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Core.V1.xml">
+    <edmx:Include Namespace="Org.OData.Core.V1" Alias="OData"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Capabilities.V1.xml">
+    <edmx:Include Namespace="Org.OData.Capabilities.V1" Alias="Capabilities"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/Resource_v1.xml">
+    <edmx:Include Namespace="Resource.v1_0_0"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/RedfishExtensions_v1.xml">
+    <edmx:Include Namespace="RedfishExtensions.v1_0_0" Alias="Redfish"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="/redfish/v1/schema/AMIPcieSwitchPort_v1.xml">
+    <edmx:Include Namespace="AMIPcieSwitchPort"/>
+    <edmx:Include Namespace="AMIPcieSwitchPort.v1_0_0"/>
+  </edmx:Reference>
+
+  <edmx:DataServices>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="AMIPcieSwitchPortCollection">
+      <Annotation Term="Redfish.OwningEntity" String="AMI"/>
+
+      <EntityType Name="AMIPcieSwitchPortCollection" BaseType="Resource.v1_0_0.ResourceCollection">
+        <Annotation Term="OData.Description" String="A Collection of  PcieSwitch Port Resource instances."/>
+        <Annotation Term="OData.LongDescription" String="This Resource shall represent a Collection of PcieSwitch Port instances for a Redfish implementation."/>
+        <Annotation Term="Capabilities.InsertRestrictions">
+          <Record>
+            <PropertyValue Property="Insertable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.UpdateRestrictions">
+          <Record>
+            <PropertyValue Property="Updatable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.DeleteRestrictions">
+          <Record>
+            <PropertyValue Property="Deletable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Redfish.Uris">
+          <Collection>
+		  <String>/redfish/v1/Oem/AMI/PcieSwitch/{PcieSwitchId}/Ports</String>
+          </Collection>
+        </Annotation>
+        <NavigationProperty Name="Members" Type="Collection(AMIPcieSwitchPort.AMIPcieSwitchPort)">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/Read"/>
+          <Annotation Term="OData.Description" String="The members of this collection."/>
+          <Annotation Term="OData.LongDescription" String="This property shall contain an array of links to the members of this collection."/>
+          <Annotation Term="OData.AutoExpandReferences"/>
+          <Annotation Term="Redfish.Required"/>
+        </NavigationProperty>
+      </EntityType>
+
+    </Schema>
+  </edmx:DataServices>
+</edmx:Edmx>
+
diff --git a/static/redfish/v1/schema/AMIPcieSwitchPort_v1.xml b/static/redfish/v1/schema/AMIPcieSwitchPort_v1.xml
new file mode 100644
index 00000000..0fdd91d1
--- /dev/null
+++ b/static/redfish/v1/schema/AMIPcieSwitchPort_v1.xml
@@ -0,0 +1,163 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!---->
+<!--################################################################################       -->
+<!--# Redfish Schema:  AMIPcieSwitchPort                                                   -->
+<!--#                                                                                      -->
+<!--################################################################################       -->
+<!---->
+<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.0">
+
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Core.V1.xml">
+    <edmx:Include Namespace="Org.OData.Core.V1" Alias="OData"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Capabilities.V1.xml">
+    <edmx:Include Namespace="Org.OData.Capabilities.V1" Alias="Capabilities"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Measures.V1.xml">
+    <edmx:Include Namespace="Org.OData.Measures.V1" Alias="Measures"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/Resource_v1.xml">
+    <edmx:Include Namespace="Resource"/>
+    <edmx:Include Namespace="Resource.v1_0_0"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/RedfishExtensions_v1.xml">
+    <edmx:Include Namespace="RedfishExtensions.v1_0_0" Alias="Redfish"/>
+    <edmx:Include Namespace="Validation.v1_0_0" Alias="Validation"/>
+  </edmx:Reference>
+
+  <edmx:DataServices>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="AMIPcieSwitchPort">
+      <Annotation Term="Redfish.OwningEntity" String="AMI"/>
+
+      <EntityType Name="AMIPcieSwitchPort" BaseType="Resource.v1_0_0.Resource" Abstract="true">
+        <Annotation Term="OData.Description" String="The PcieSwitch shema contains PcieSwitch Port detail."/>
+        <Annotation Term="OData.LongDescription" String="The PcieSwitch shema contains PcieSwitch Port detail."/>
+        <Annotation Term="Capabilities.InsertRestrictions">
+          <Record>
+            <PropertyValue Property="Insertable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.UpdateRestrictions">
+          <Record>
+            <PropertyValue Property="Updatable" Bool="true"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.DeleteRestrictions">
+          <Record>
+            <PropertyValue Property="Deletable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Redfish.Uris">
+          <Collection>
+		  <String>/redfish/v1/Oem/AMI/PcieSwitch/{PcieSwitchId}/Ports/{PcieSwitchPortId}</String>
+          </Collection>
+        </Annotation>
+      </EntityType>
+    </Schema>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="AMIPcieSwitchPort.v1_0_0">
+      <Annotation Term="Redfish.OwningEntity" String="AMI"/>
+      <Annotation Term="Redfish.Release" String="1.0"/>
+
+      <EntityType Name="AMIPcieSwitchPort" BaseType="AMIPcieSwitchPort.AMIPcieSwitchPort">
+        <Property Name="ActiveLanes" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="This property indicates ActiveLanes."/>
+          <Annotation Term="OData.LongDescription" String="This property indicates ActiveLanes."/>
+        </Property>
+        <Property Name="BusPrimary" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="This property indicates BusPrimary."/>
+          <Annotation Term="OData.LongDescription" String="This property indicates BusPrimary."/>
+        </Property>
+        <Property Name="BusSecondary" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Description" String="This property indicates BusSecondary."/>
+          <Annotation Term="OData.LongDescription" String="This property indicates BusSecondary."/>
+        </Property>
+        <Property Name="BusSubordinate" Type="Edm.Decimal">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/Read"/>
+          <Annotation Term="OData.Description" String="This property indicates BusSubordinate."/>
+          <Annotation Term="OData.LongDescription" String="This property indicates BusSubordinate."/>
+        </Property>
+        <Property Name="GID" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="This property indicates GID."/>
+          <Annotation Term="OData.LongDescription" String="This property indicates GID."/>
+        </Property>
+        <Property Name="LinkModeType" Type="Edm.Decimal">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates LinkModeType."/>
+          <Annotation Term="OData.LongDescription" String="It indicates LinkModeType."/>
+        </Property>
+        <Property Name="MaxLinkSpeed" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates MaxLinkSpeed."/>
+          <Annotation Term="OData.LongDescription" String="It indicates MaxLinkSpeed."/>
+        </Property>
+        <Property Name="MaxLinkWidth" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates MaxLinkWidth."/>
+          <Annotation Term="OData.LongDescription" String="It indicates MaxLinkWidth."/>
+        </Property>
+        <Property Name="MaxPayloadSize" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates MaxPayloadSize."/>
+          <Annotation Term="OData.LongDescription" String="It indicates MaxPayloadSize."/>
+        </Property>
+        <Property Name="MaxPayloadSizeSupport" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates MaxPayloadSizeSupport."/>
+          <Annotation Term="OData.LongDescription" String="It indicates MaxPayloadSizeSupport."/>
+        </Property>
+        <Property Name="MaxReadRequestSize" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates MaxReadRequestSize."/>
+          <Annotation Term="OData.LongDescription" String="It indicates MaxReadRequestSize."/>
+        </Property>
+        <Property Name="Name" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Name of a Port."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Name of a Port."/>
+        </Property>
+        <Property Name="NegotiatedLinkSpeed" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates NegotiatedLinkSpeed."/>
+          <Annotation Term="OData.LongDescription" String="It indicates NegotiatedLinkSpeed."/>
+        </Property>
+        <Property Name="NegotiatedLinkWidth" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates NegotiatedLinkWidth."/>
+          <Annotation Term="OData.LongDescription" String="It indicates NegotiatedLinkWidth."/>
+        </Property>
+        <Property Name="PortNumber" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PortNumber."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PortNumber."/>
+        </Property>
+        <Property Name="PortType" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PortType."/>
+	  <Annotation Term="OData.LongDescription" String="It indicates PortType."/>
+        </Property>
+        <Property Name="Reserved" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+	  <Annotation Term="OData.Description" String="It indicates Reserved status."/>
+	  <Annotation Term="OData.LongDescription" String="It indicates Reserved status"/>
+        </Property>
+        <Property Name="Reserved1" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Reserved1 status."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Reserved1 status."/>
+        </Property>
+        <Property Name="isEnabled" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Port is Enabled or not."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Port is Enabled or not."/>
+        </Property>
+      </EntityType>
+    </Schema>
+
+  </edmx:DataServices>
+</edmx:Edmx>
+
diff --git a/static/redfish/v1/schema/AMIPcieSwitch_v1.xml b/static/redfish/v1/schema/AMIPcieSwitch_v1.xml
new file mode 100644
index 00000000..878c21d5
--- /dev/null
+++ b/static/redfish/v1/schema/AMIPcieSwitch_v1.xml
@@ -0,0 +1,222 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!---->
+<!--################################################################################       -->
+<!--# Redfish Schema:  AMIPcieSwitch                                                       -->
+<!--#                                                                                      -->
+<!--################################################################################       -->
+<!---->
+<edmx:Edmx xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" Version="4.0">
+
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Core.V1.xml">
+    <edmx:Include Namespace="Org.OData.Core.V1" Alias="OData"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Capabilities.V1.xml">
+    <edmx:Include Namespace="Org.OData.Capabilities.V1" Alias="Capabilities"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/errata03/csd01/complete/vocabularies/Org.OData.Measures.V1.xml">
+    <edmx:Include Namespace="Org.OData.Measures.V1" Alias="Measures"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/Resource_v1.xml">
+    <edmx:Include Namespace="Resource"/>
+    <edmx:Include Namespace="Resource.v1_0_0"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/RedfishExtensions_v1.xml">
+    <edmx:Include Namespace="RedfishExtensions.v1_0_0" Alias="Redfish"/>
+    <edmx:Include Namespace="Validation.v1_0_0" Alias="Validation"/>
+  </edmx:Reference>
+  <edmx:Reference Uri="/redfish/v1/schema/AMIPcieSwitchPortCollection_v1.xml">
+    <edmx:Include Namespace="AMIPcieSwitchPortCollection"/>
+  </edmx:Reference>
+
+  <edmx:DataServices>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="AMIPcieSwitch">
+      <Annotation Term="Redfish.OwningEntity" String="AMI"/>
+
+      <EntityType Name="AMIPcieSwitch" BaseType="Resource.v1_0_0.Resource" Abstract="true">
+        <Annotation Term="OData.Description" String="The AMIPcieSwitch schema contains properties that describe a device details."/>
+        <Annotation Term="OData.LongDescription" String="This resource contains a simple PcieSwitch for a Redfish implementation.."/>
+        <Annotation Term="Capabilities.InsertRestrictions">
+          <Record>
+            <PropertyValue Property="Insertable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.UpdateRestrictions">
+          <Record>
+            <PropertyValue Property="Updatable" Bool="true"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Capabilities.DeleteRestrictions">
+          <Record>
+            <PropertyValue Property="Deletable" Bool="false"/>
+          </Record>
+        </Annotation>
+        <Annotation Term="Redfish.Uris">
+          <Collection>
+		  <String>/redfish/v1/Oem/AMI/PcieSwitch/{PcieSwitchId}</String>
+          </Collection>
+        </Annotation>
+      </EntityType>
+    </Schema>
+
+    <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="AMIPcieSwitch.v1_0_0">
+      <Annotation Term="Redfish.OwningEntity" String="AMI"/>
+      <Annotation Term="Redfish.Release" String="1.0"/>
+
+      <EntityType Name="AMIPcieSwitch" BaseType="AMIPcieSwitch.AMIPcieSwitch">
+        <Property Name="ComponentId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="Represents ComponentId."/>
+          <Annotation Term="OData.LongDescription" String="Represents ComponentId of an intance."/>
+        </Property>
+        <Property Name="EnclosureWwidHigh" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates EnclosureWwidHigh."/>
+          <Annotation Term="OData.LongDescription" String="It indicates EnclosureWwidHigh."/>
+        </Property>
+        <Property Name="EnclosureWwidLow" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Description" String="It indicates EnclosureWwidLow."/>
+          <Annotation Term="OData.LongDescription" String="It indicates EnclosureWwidLow."/>
+        </Property>
+        <Property Name="FwVersion" Type="Collection(Edm.Decimal)">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/Read"/>
+          <Annotation Term="OData.Description" String="It indicates FwVersion."/>
+          <Annotation Term="OData.LongDescription" String="It indicates FwVersions."/>
+        </Property>
+        <Property Name="HandleType" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates HandleType."/>
+          <Annotation Term="OData.LongDescription" String="It indicates HandleType."/>
+        </Property>
+        <Property Name="LastOp" Type="Edm.String">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates LastOp."/>
+          <Annotation Term="OData.LongDescription" String="It indicates LastOp."/>
+        </Property>
+        <Property Name="LastOpStatus" Type="Edm.String" Nullable="false">
+	  <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates LastOpStatus."/>
+          <Annotation Term="OData.LongDescription" String="It indicates LastOpStatus."/>
+        </Property>
+        <Property Name="Manufacturer" Type="Edm.String" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Manufacturer."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Manufacturer."/>
+        </Property>
+        <Property Name="Mode" Type="Edm.String" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Mode."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Mode."/>
+        </Property>
+        <Property Name="Model" Type="Edm.String" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Model."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Model."/>
+        </Property>
+        <Property Name="Name" Type="Edm.String" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Name."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Name."/>
+        </Property>
+        <Property Name="NumLanes" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates NumLanes."/>
+          <Annotation Term="OData.LongDescription" String="It indicates NumLanes."/>
+        </Property>
+        <Property Name="PciAddrBusNumber" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciAddrBusNumber."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciAddrBusNumber."/>
+        </Property>
+        <Property Name="PciAddrDeviceNumber" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciAddrDeviceNumber."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciAddrDeviceNumber."/>
+        </Property>
+        <Property Name="PciAddrFunctionNumber" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciAddrFunctionNumber."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciAddrFunctionNumber."/>
+        </Property>
+        <Property Name="PciDeviceId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciDeviceId."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciDeviceId."/>
+        </Property>
+        <Property Name="PciMptDeviceId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciMptDeviceId."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciMptDeviceId."/>
+        </Property>
+        <Property Name="PciSSDeviceId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciSSDeviceId."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciSSDeviceId."/>
+        </Property>
+        <Property Name="PciSSVendorId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciSSVendorId."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciSSVendorId."/>
+        </Property>
+        <Property Name="PciVendorId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates PciVendorId."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PciVendorId."/>
+        </Property>
+        <NavigationProperty Name="Port" Type="AMIPcieSwitchPortCollection.AMIPcieSwitchPortCollection" ContainsTarget="true">
+          <Annotation Term="OData.Description" String="The link to the collection of PcieSwitchPortCollection."/>
+          <Annotation Term="OData.LongDescription" String="This property shall contain a link to the collection of PcieSwitchPortCollection."/>
+          <Annotation Term="OData.AutoExpandReferences"/>
+        </NavigationProperty>
+        <Property Name="Ports" Type="Collection(Edm.Decimal)" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+	  <Annotation Term="OData.Description" String="It indicates PcieSwitch Ports."/>
+          <Annotation Term="OData.LongDescription" String="It indicates PcieSwitch Ports."/>
+        </Property>
+        <Property Name="ProductHandle" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates ProductHandle."/>
+          <Annotation Term="OData.LongDescription" String="It indicates ProductHandle."/>
+        </Property>
+        <Property Name="RevisionId" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates RevisionId."/>
+          <Annotation Term="OData.LongDescription" String="It indicates RevisionId."/>
+        </Property>
+        <Property Name="SASAddressHigh" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates SASAddressHigh."/>
+          <Annotation Term="OData.LongDescription" String="It indicates SASAddressHigh."/>
+        </Property>
+        <Property Name="SASAddressLow" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates SASAddressLow."/>
+          <Annotation Term="OData.LongDescription" String="It indicates SASAddressLow."/>
+        </Property>
+        <Property Name="SbrVersion" Type="Collection(Edm.Decimal)" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates SbrVersion."/>
+	  <Annotation Term="OData.LongDescription" String="It indicates SbrVersion."/>
+        </Property>
+        <Property Name="SmApiVersion" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates SmApiVersion."/>
+          <Annotation Term="OData.LongDescription" String="It indicates SmApiVersion.."/>
+        </Property>
+        <Property Name="SwitchMode" Type="Edm.Decimal" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates SwitchMode."/>
+          <Annotation Term="OData.LongDescription" String="It indicates SwitchMode."/>
+        </Property>
+        <Property Name="Temperature" Type="Edm.String" Nullable="false">
+          <Annotation Term="OData.Permissions" EnumMember="OData.Permission/ReadWrite"/>
+          <Annotation Term="OData.Description" String="It indicates Temperature."/>
+          <Annotation Term="OData.LongDescription" String="It indicates Temperature."/>
+        </Property>
+      </EntityType>
+    </Schema>
+
+  </edmx:DataServices>
+</edmx:Edmx>
+~
+
diff --git a/static/redfish/v1/schema/OemServiceRoot_v1.xml b/static/redfish/v1/schema/OemServiceRoot_v1.xml
index 490efe2f..47cf7625 100644
--- a/static/redfish/v1/schema/OemServiceRoot_v1.xml
+++ b/static/redfish/v1/schema/OemServiceRoot_v1.xml
@@ -19,6 +19,11 @@
     <edmx:Include Namespace="CupsService"/>
   </edmx:Reference>
 
+   <edmx:Reference Uri="http://redfish.dmtf.org/schemas/v1/AMIPcieSwitchCollection_v1.xml">
+      <edmx:Include Namespace="AMIPcieSwitchCollection"/>
+      <edmx:Include Namespace="AMIPcieSwitchCollection.v1_0_0"/>
+   </edmx:Reference>
+
   <edmx:DataServices>
 
     <Schema xmlns="http://docs.oasis-open.org/odata/ns/edm" Namespace="OemServiceRoot">
@@ -33,7 +38,8 @@
           <Annotation Term="OData.AdditionalProperties" Bool="true"/>
           <Annotation Term="OData.Description" String="OemServiceRoot Oem properties."/>
           <Annotation Term="OData.AutoExpand"/>
-          <Property Name="Intel" Type="OemServiceRoot.v1_0_0.Intel"/>
+	  <Property Name="Intel" Type="OemServiceRoot.v1_0_0.Intel"/>
+	  <property Name="AMI" Type="OemServiceRoot.v1_0_0.AMI"/>
         </ComplexType>
 
         <ComplexType Name="Intel" BaseType="Resource.OemObject">
@@ -47,6 +53,19 @@
             <Annotation Term="OData.AutoExpandReferences"/>
           </NavigationProperty>
         </ComplexType>
+
+        <ComplexType Name="AMI" BaseType="Resource.OemObject">
+          <Annotation Term="OData.AdditionalProperties" Bool="true"/>
+          <Annotation Term="OData.Description" String="Oem properties for AMI."/>
+          <Annotation Term="OData.AutoExpand"/>
+          <NavigationProperty Name="PcieSwitch" Type="AMIPcieSwitchCollection.AMIPcieSwitchCollection" Nullable="false">
+            <Annotation Term="OData.Permissions" EnumMember="OData.Permission/Read"/>
+            <Annotation Term="OData.Description" String="This allows to link the redfish PcieSwitch collection."/>
+            <Annotation Term="OData.LongDescription" String="This allows to link the redfish PcieSwitch collection."/>
+            <Annotation Term="OData.AutoExpandReferences"/>
+           </NavigationProperty>
+         </ComplexType>
+
     </Schema>
 
   </edmx:DataServices>
-- 
2.34.1

